modules/wallet/scan.go

```diff
@@ -0,0 +1,198 @@
+package wallet
+
+import (
+	"fmt"
+
+	"github.com/NebulousLabs/Sia/build"
+	"github.com/NebulousLabs/Sia/modules"
+	"github.com/NebulousLabs/Sia/types"
+)
+
+// numInitialKeys is the number of keys generated by the seedScanner before
+// scanning the blockchain for the first time.
+var numInitialKeys = func() uint64 {
+	switch build.Release {
+	case "dev":
+		return 1e4
+	case "standard":
+		return 1e6
+	case "testing":
+		return 1e3
+	default:
+		panic("unrecognized build.Release")
+	}
+}()
+
+// maxScanKeys is the number of maximum number of keys the seedScanner will
+// generate before giving up.
+var maxScanKeys = func() uint64 {
+	switch build.Release {
+	case "dev":
+		return 1e6
+	case "standard":
+		return 100e6
+	case "testing":
+		return 1e5
+	default:
+		panic("unrecognized build.Release")
+	}
+}()
+
+var errMaxKeys = fmt.Errorf("refused to generate more than %v keys from seed", maxScanKeys)
+
+type scannedOutput struct {
+	id        types.OutputID
+	value     types.Currency
+	seedIndex uint64
+}
+
+// A seedScanner scans the blockchain for addresses that belong to a given
+// seed.
+type seedScanner struct {
+	blockheight      types.BlockHeight
+	dustThreshold    types.Currency              // minimum value of outputs to be included
+	keys             map[types.UnlockHash]uint64 // map address to seed index
+	largestIndexSeen uint64                      // largest index that has appeared in the blockchain
+	minerOutputs     map[types.BlockHeight][]scannedOutput
+	seed             modules.Seed
+	siacoinOutputs   map[types.OutputID]scannedOutput
+}
+
+func (s *seedScanner) numKeys() uint64 {
+	return uint64(len(s.keys))
+}
+
+// generateKeys generates n additional keys from the seedScanner's seed.
+func (s *seedScanner) generateKeys(n uint64) {
+	initialProgress := s.numKeys()
+	for i, k := range generateKeys(s.seed, initialProgress, n) {
+		s.keys[k.UnlockConditions.UnlockHash()] = initialProgress + uint64(i)
+	}
+}
+
+// ProcessConsensusChange scans the blockchain for information relevant to the
+// seedScanner.
+func (s *seedScanner) ProcessConsensusChange(cc modules.ConsensusChange) {
+	s.blockheight -= types.BlockHeight(len(cc.RevertedBlocks))
+	// update outputs
```

Comment:
aren't you going to have problems if you save an output that got reverted and then double-spent? A lot of user wallets are going to have double spends in them because the wallet for a long time would not remember which outputs it had already used if you restarted it. In fact I'm not even sure 1.0.0 solved that issue.